### ISSUE TITLE
Wrong `mut &mut self` suggestion

### ISSUE THREAD

**Issue by DenisKolodin** [rust-lang/rust#34126]
I've found wrong suggestion with the following code ([play](https://is.gd/YgknlV)):

``` rust
struct Z { }

impl Z {
  fn run(&self, z: &mut Z) { }
  fn start(&mut self) { self.run(&mut self); }
}

fn main() {
  let mut z = Z {};
  z.start();
}
```

stable release prints like in #31424:

``` rust
<anon>:8:19: 8:23 error: cannot borrow immutable argument `self` as mutable
<anon>:8     self.run(&mut self);
                           ^~~~
<anon>:7:17: 7:21 help: to make the argument mutable, use `mut` as shown:
<anon>:    fn start(&mut mut self) {
error: aborting due to previous error
playpen: application terminated with error code 101
```

but nightly takes suggestion to use `mut &mut self`:

``` rust
error: cannot borrow immutable argument `self` as mutable
 --> <anon>:8:19
7 |>   fn start(&mut self) {
  |>            --------- use `mut &mut self` here to make mutable
8 |>     self.run(&mut self);
  |>                   ^^^^ cannot borrow mutably

error: aborting due to previous error
playpen: application terminated with error code 101
```

**Comment by durka** [comment]
cc @birkenfeld

**Comment by birkenfeld** [comment]
This is due to the change in how self arguments are represented. Now we can probably properly check, I'll have a look.

**Comment by alexandermerritt** [comment]
Possibly related? #34337

**Comment by nikomatsakis** [comment]
Fixed by #37531 it seems
